rsearch/resources.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2016 Pani Networks
+// All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package rsearch
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+)
+
+/*
+{"type":"ADDED","object":{"kind":"Namespace","apiVersion":"v1","metadata":{"name":"default","selfLink":"/api/v1/namespaces/default","uid":"d10db271-dc03-11e5-9c86-0213e1312dc5","resourceVersion":"6","creationTimestamp":"2016-02-25T21:07:45Z"},"spec":{"finalizers":["kubernetes"]},"status":{"phase":"Active"}}}
+*/
+
+type Event struct {
+	Type   string     `json:"Type"`
+	Object KubeObject `json:"object"`
+}
+
+type KubeObject struct {
+	Kind       string            `json:"kind"`
+	Spec       Spec              `json:"spec"`
+	ApiVersion string            `json:"apiVersion"`
+	Metadata   Metadata          `json:"metadata"`
+	Status     map[string]string `json:"status,omitempty"`
+}
+
+func (o KubeObject) makeId() string {
+	id := o.Metadata.Name + "/" + o.Metadata.Namespace
+	return id
+}
+
+func (o KubeObject) getSelector(config Config) string {
+	var selector string
+	// TODO this should use Config.Resource.Selector path instead of podSelector
+	for k, v := range o.Spec.PodSelector {
+		selector = k + "/" + v + "#"
+	}
+	return selector
+}
+
+// TODO need to find a way to use different specs for different resources
+type Spec struct {
+	AllowIncoming map[string]interface{} `json:"allowIncoming"`
+	PodSelector   map[string]string      `json:"podSelector"`
+}
+
+type Metadata struct {
+	Name              string            `json:"name"`
+	Namespace         string            `json:"namespace"`
+	SelfLink          string            `json:"selfLink"`
+	Uid               string            `json:"uid"`
+	ResourceVersion   string            `json:"resourceVersion"`
+	CreationTimestamp string            `json:"creationTimestamp"`
+	Labels            map[string]string `json:"labels"`
+}
+
+func watchEvents(done <-chan Done, url string, config Config, resp *http.Response, out chan Event) {
+	if config.Server.Debug {
+		log.Println("Received namespace related event from kubernetes", resp.Body)
+	}
+
+	tick := time.Tick(5 * time.Second)
+	dec := json.NewDecoder(resp.Body)
+	var e Event
+
+	for {
+		select {
+		case <-tick:
+			// If decoded successfully
+			if err := dec.Decode(&e); err != nil {
```

Comment:
Right, that was an effort to handle GET timeouts. Apparently, i overlooked damage i've done while fixing it.